uamqp/client.py

```diff
@@ -639,14 +638,30 @@ def receive_message_batch(self, max_batch_size=None, on_message_received=None, t
         else:
             return batch
 
-        while receiving and not expired and len(batch) < max_batch_size:
-            receiving = self.do_work(batch=max_batch_size)
-            while len(batch) < max_batch_size:
-                try:
-                    batch.append(self._received_messages.get_nowait())
-                    self._received_messages.task_done()
-                except queue.Empty:
-                    break
+        to_receive_size = max_batch_size - len(batch)
+        before_queue_size = self._received_messages.qsize()
+
+        while receiving and to_receive_size > 0:
             if timeout and time.time() > timeout:
-                expired = True
+                break
+
+            receiving = self.do_work(batch=to_receive_size)
```

Comment:
Remind me - what does specifying `batch=` here achieve? Does this mean we will continue polling the socket for _exactly_ `to_receive_size` number of messages? Or 'best effort'?
I'm wondering if we're duplicating logic - for example if `do_work` is already going to return us exactly the number of events we want, or otherwise timeout, then that's also the logic that this function is implementing. But I can't remember exactly what the `batch` flag does....

Reply:
it ended up being called here: https://github.com/Azure/azure-uamqp-python/blob/pyproto/uamqp/_connection.py#L535-L560

The `batch` is for "The number of frames to attempt to read and process before returning" and depending on the implementation, it's doing `best effort` -- will try to consume as much as `batch` frames, but if there's no more frame coming in, the for loop would break.

Reply:
Cool okay. I feel like we could shuffle this to be more efficient - but this follows the existing logic and it's working so we'll go with it for now :)